components/director/internal/formationmapping/auth_middleware.go

```diff
@@ -0,0 +1,317 @@
+				respondWithError(ctx, w, statusCode, errors.New("An unexpected error occurred while processing the request"))
+				return
+			}
+
+			if !isAuthorized {
+				httputils.Respond(w, http.StatusUnauthorized)
+				return
+			}
+
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+// isAuthorized verify through custom logic the caller is authorized to update the formation assignment status
+func (a *Authenticator) isAuthorized(ctx context.Context, formationAssignmentID string) (bool, int, error) {
+	consumerInfo, err := consumer.LoadFromContext(ctx)
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrap(err, "while fetching consumer info from context")
+	}
+	consumerID := consumerInfo.ConsumerID
+	consumerType := consumerInfo.ConsumerType
+
+	tx, err := a.transact.Begin()
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrap(err, "Unable to establish connection with database")
+	}
+	defer a.transact.RollbackUnlessCommitted(ctx, tx)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	fa, err := a.faService.GetGlobalByID(ctx, formationAssignmentID)
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrapf(err, "while getting formation assignment with ID: %q globally", formationAssignmentID)
+	}
+
+	consumerTenantPair, err := tenant.LoadTenantPairFromContext(ctx)
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrap(err, "while loading tenant pair from context")
+	}
+	consumerInternalTenantID := consumerTenantPair.InternalID
+	consumerExternalTenantID := consumerTenantPair.ExternalID
+
+	log.C(ctx).Infof("Consumer with internal ID: %q and external ID: %q with type: %q is trying to update formation assignment with ID: %q for formation with ID: %q about source: %q and source type: %q, and target: %q and target type: %q", consumerInternalTenantID, consumerExternalTenantID, consumerType, fa.ID, fa.FormationID, fa.Source, fa.SourceType, fa.Target, fa.TargetType)
+	if fa.TargetType == model.FormationAssignmentTypeApplication {
+		log.C(ctx).Infof("The formation assignment that is being updated has type: %s and ID: %q", model.FormationAssignmentTypeApplication, fa.Target)
+
+		app, err := a.appRepo.GetGlobalByID(ctx, fa.Target)
+		if err != nil {
+			return false, http.StatusInternalServerError, errors.Wrapf(err, "while getting application with ID: %q globally", fa.Target)
+		}
+		log.C(ctx).Infof("Successfully got application with ID: %q globally", fa.Target)
+
+		// If the consumer is integration system validate the formation assignment type is application that can be managed by the integration system caller
+		if consumerType == consumer.IntegrationSystem && app.IntegrationSystemID != nil && *app.IntegrationSystemID == consumerID {
+			log.C(ctx).Infof("The caller with ID: %q and type: %q has owner access to the formation assignment with target: %q and target type: %q that is being updated", consumerID, consumerType, fa.Target, fa.TargetType)
```

Comment:
I was wondering what will be the best way to log these operations because as you mentioned in the other comments the caller could be diffent, that's why on all of them additionally to the caller ID we have the type of it.